draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+is often desirable to use this result. One way to do this would be to wait until
+the promise has been resolved and then send the next message to the result.
+While this approach can be taken within OCapN, promise pipelining is preferred.
+When a message is sent with an `answer-pos` specified, the promise on the remote
+session can then be referenced and messaged with the
+[`desc:answer`](#desc-answer) descriptor. When the result is ready, messages
+sent to this promise will have them delivered to the result.
+
+If the answer contains multiple values, the [`op:select`](#op-select) operator
+MUST be used to select the value to be pipelined on. This works by exporting the
+individual value in another answer position, which then can be used to message
+the promise. If multiple values are pipelined on, without the select, the
+resulting promise MUST be broken with an error and not fulfilled.
+
+### Promise pipelining by example
+
+In this example we send an initial message to the object exported at position
+`5` (a factory builder) and tell the remote session to export the promise at
+answer position `3` (a car factory), then we send a message to the promise at
+that answer position to make a car and export it at answer position `4`, finally
+we send a message to the promise of a car at answer position `4` to drive it.
+
+The messages delivered are as follows:
+
+```
+<op:deliver <desc:export 5> ['make-car-factory] 3 <desc:import-object 15>>
+<op:deliver <desc:answer 3> ['make-car] 4 <desc:import-object 16>>
+<op:deliver <desc:answer 4> ['drive] 5 <desc:import-object 17>>
+```
+
+In this way, it is not necessary to wait to dispatch messages until the promise
+has been resolved. Messages can just be sent in advance to a promise of the
+eventual object as they normally would to its eventual resolution. This also
+improves efficiency by reducing the number of round trips needed to perform the
+same task. Take the above example of a car factory which produces a car, the car
+then can be driven which will produce a noise. Without promise pipelining this
+would look something like this:
+
+1.  Send message to create car factory
+2.  Get back a reference to the car factory once it is made
+3.  Send message to car factory to make a car
+4.  Get back a reference to the car
+5.  Send message to car to drive
+6.  Get back the driving noise.
+
+This is compared to when you use promise pipelining:
+
+1.  Send message to create car factory
+2.  Send message to car factory promise to make a car
+3.  Send message to car promise to drive
+4.  Get back driving noise
+
+The other CapTP session, which is making the car factory, car, and finally car
+noise, is able deliver those messages immediately to the local objects on their
+side as they become fulfilled and send the car noise back at the end.
+
+
+# [Cryptography](#cryptography)
+
+Each CapTP session has a per session key pair which is used for signing certain
```

Comment:
Sometimes it is "key pair", like here. Other times it is "key-pair". Could one of those variants be used for consistency?